README.md
[![Build Status](https://travis-ci.org/stefanvignir/iceland_gem.svg)](https://travis-ci.org/stefanvignir/iceland_gem)
[![Test Coverage](https://codeclimate.com/github/stefanvignir/iceland_gem/badges/coverage.svg)](https://codeclimate.com/github/stefanvignir/iceland_gem/coverage)
[![Code Climate](https://codeclimate.com/github/stefanvignir/iceland_gem/badges/gpa.svg)](https://codeclimate.com/github/stefanvignir/iceland_gem)

# The Iceland Gem

The Iceland Gem handles Icelandic "kennitala" personal/entity identifiers and Icelandic postal codes.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'iceland'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install iceland

## Usage

### Postal Codes

Icelandic postal codes are 3-digit numeric identifiers, with the first digit indicating a region and the rest identifying a specific locale within the region.

Names of locales are provided in dative form by default as per Icelandic postal convention, but nomative forms can be returned by setting the `force_nominative` parameter to `true` when using the `Iceland.all_postal_codes` and `Iceland.locale_by_postal_code` methods.

(Note that we use the term "locale" as per Universal Postal Union convention to name the town, city or other location the postal code is assigned to.)

#### Examples

```ruby
# Get all postal codes, skipping codes assigned to P.O. boxes.
# This is useful when generating forms.
Iceland::PostalCode.list
# => [{:postal_code=>101, :locale=>"Reykjavík"}, {:postal_code=>103, [...]

# Get the same postal codes in nomative form
Iceland::PostalCode.list(nominative: true)
# => [{:postal_code=>101, :locale=>"Reykjavík"}, {:postal_code=>103, [...]

# Get all postal codes, including P.O. boxes
Iceland::PostalCode.list(include_po_boxes: true)
# => [{:postal_code=>101, :locale=>"Reykjavík"}, {:postal_code=>103, [...]

# Get the name of locale based on its postal code.
# This is useful when you only save the postal code, but not the locale in a database.
Iceland::PostalCode.find_locale(311)
# => "Borgarnesi (dreifbýli)"

# Get the name of the locale in nominative form
Iceland.find_locale(postal_code: 311, nominative: true)
# => "Borgarnes (dreifbýli)"
```

### The Kennitala Class

The Iceland Gem provides a class to handle "kennitala" identifier codes. The class can be used to sanitize the identifiers and read information like the date of birth (or date of registration in the case of companies and organization), age and the type of entity.

The class does not access external APIs or databases such National Registry or the Company Registry, so names and status (sex/gender, death, bankruptcy, credit rating etc.) cannot be accessed using the class. However, it can be used to sanitize and validate such data before being sent to external APIs, as such services are provided by private companies, which often charge a specific amount for each query.

#### Uses of kennitala

Unlike the US Social Security number and equivalents, the kennitala is only used for identification of persons and companies (as well as other registered organizations) — and is often used internally by educational institutions, companies and other organization as a primary identifier for persons (e.g. school, employee, customer and frequent flyer ID). It is not to be used for authentication (i.e. a password) and is not considered a secret per se. While a kennitala can be kept unencrypted in a database, publishing a kennitala or a list of them is generally not considered good practice and might cause liability.

A kennitala is assigned to every newborn person and foreign nationals residing in Iceland as well as organizations and companies operating there. It is statically assigned and can not be changed.

Article II, paragraph 10 of the 77/2000 Act on Data Protection (http://www.althingi.is/lagas/nuna/2000077.html) provides the legal framework regarding the use and processing of the kennitala in Iceland:

> The use of a kennitala is allowed if it has a an objective cause and is necessary to ensure reliable identification of persons. The Data Protection Authority may ban or order the use of kennitala.

#### Technicalities

The kennitala (`DDMMYY-RRCM`) is a 10-digit numeric string consisting on a date (date of birth for persons, date of registration for companies) in the form of `DDMMYY`, three two random digits (`RR`) a check digit (`C`) and a century identifier (`M`). A hyphen or space is often added between the year and random values (Example: `010130-2989`).

The number 40 is added to the registration day of companies and organizations. Hence, a kennitala for a company registered at January 1 1990 starts with `410190` as opposed to `010190` for a person born that day.

The century identifier has 3 legal values. `8` for the 19th century, `9` for the 20th century and `0` for the 21st century.

#### Examples

##### Working with Kennitala objects

```ruby
# Initialize a Kennitala object.
# The string provided may include spaces, hyphens and alphabetical characters,
# which will then be erased from the resulting string.
k = Iceland::Kennitala.new(' 010130-2989')
# => #<Kennitala:0x007fe35d041bc0 @value="0101302989">

# Invalid strings are rejected with an argument error
f = Iceland::Kennitala.new('010130-2979')
# => ArgumentError: Kennitala is invalid

# If no kennitala string is specified, a random one will be generated
r = Iceland::Kennitala.new
# => #<Kennitala:0x007fc589339f18 @value="2009155509">

# Retrieve the kennitala as a string.
# This is a sanitized string, without any non-numeric characters.
# Pretty useful when storing it in a database.
k.to_s
# => "0101302989"

# Pretty print the kennitala
# Adds a space between the 6th and the 7th digits for readability
k.pp
# => "010130 2989"

# You can also pass a string to .pp to use as a spacer
k.pp('–')
# => "010130-2989"

# You can also pass a cat to the .pp method
k.pp('🐈')
# => "010130🐈2989"

# Get the entity type (results in 'person' or 'company')
k.entity_type
# => "person"

# It's also possible to use .company and .person to achieve the same thing
k.company?
# => false

k.person?
# => true

# Cast the kennitala to a Date object
k.to_date
# => #<Date: 1930-01-01 ((2425978j,0s,0n),+0s,2299161j)>

# Get the current age of the entity. Useful for age restrictions.
k.age
# => 86
```

##### Casting strings

```ruby
# Casting a string to a Kennitala object
'010130 2989'.to_kt
# => #<Kennitala:0x007fc5893286a0 @value="0101302989">

# Get the current age based on a String
'0101302989'.to_kt.age
# => 86
```

## Todo

* Split the gem into modules and make the this gem a collection gem, depending on and thus including all the others
* Initialization ob objects and instance methods for the `Iceland::PostalCode` class
* Add Administrative Divisions

## About the data

The Postal code data is based on data files provided by Iceland Post. The files are available at http://www.postur.is/um-postinn/posthus/postnumer/gagnaskrar/ and are provided for free for any use by individuals and organizations.

Strings are expected to be UTF-8.

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/stefanvignir/iceland_gem.

Do make sure that the `rspec` unit tests run before sending a pull request and write tests for any new functionality you add. Also run `rubocop` to check if your code adheres to the Ruby Style Guide and other conventions.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
## Version 0.2

### API-level changes

* Calling postcode related methods directly from the Iceland module, for example `Iceland.all_postal_codes` has been depreciated. Use the `Iceland::PostalCodes` class instead.
* The kennitala class has been replaced with the more appropriately namespaced `Iceland::Kennitala` class. There is no deprecation involved.

### Development dependencies

* Updating all development dependencies to their newest version.
* `codeclimate-test-reporter` is among breaking releases, requiring changes to the `spec_helper`.
* Removing overcommit as it is fucking annoying

### Tests

* Splitting specs into different files by functionality.

### Minor changes

* Updating specs and documentation to reflect the depreciation of `Fixnum` in ruby 2.4.


lib/iceland/iceland_kennitala.rb
# frozen_string_literal: true

module Iceland
  # The kennitala class
  class Kennitala
    # Initialize a Kennitala object
    #
    # @note The string provided may include spaces, hyphens and alphabetical
    #       characters, which will then be erased from the resulting string.
    #
    # @param [String, Boolean] kt_string Either Kennitala String or false.
    # @param [Boolean] is_company Wether the randomly generated kennitala is
    #                  for a company
    # @return [Kennitala] description of returned object
    #
    # @example Initialize a Kennitala object based on an unsanitized String
    #          Kennitala.new(' 010130-2989')
    #          => #<Kennitala:0x007fe35d041bc0 @value="0101302989">
    #
    # @example Invalid strings are rejected with an argument error
    #          Kennitala.new('010130-2979')
    #          => ArgumentError: Kennitala is invalid
    #
    # @example Generate a random Kennitala
    #          Kennitala.new
    #          => #<Kennitala:0x007fc589339f18 @value="2009155509">
    def initialize(kt_string = false, is_company = false)
      kt_string = fake_kt_string(is_company) if kt_string == false
      unless kt_string.class == String
        raise ArgumentError, 'Kennitala needs to be provided as a String or '\
                             'Boolean (false)'
      end
      sanitised_kt = sanitize(kt_string)
      raise ArgumentError, 'Kennitala is invalid' if sanitised_kt.nil?
      @value = sanitised_kt
    end

    # Get the type of entity - If it is a person or an organization
    #
    # @return [String] Either 'person' or 'company'
    #
    # @example Get the entity type (results in 'person' or 'company')
    #          k = Kennitala.new('0101302989')
    #          => #<Kennitala:0x007fe35d041bc0 @value="0101302989">
    #          k.entity_type
    #          => "person"
    def entity_type
      date_integer = @value[0, 2].to_i
      return 'person' if date_integer < 32
      return 'company' if (date_integer > 40) && (date_integer < 72)
    end

    # Check if the entity is a company
    #
    # @return [Boolean]
    def company?
      date_integer = @value[0, 2].to_i
      return true if (date_integer > 40) && (date_integer < 72)
      false
    end

    # Check if the entity is a person
    #
    # @return [Boolean]
    def person?
      date_integer = @value[0, 2].to_i
      return true if date_integer < 32
      false
    end

    # Get the year of birth or registration
    #
    # @return [Integer]
    def year
      century = (10 + @value[9].to_i) * 100
      year = @value[4, 2].to_i
      return century + year if (1800..1900).cover?(century)
      return 2000 + year if century == 1000
    end

    # Get the day of the month of birth or registration
    #
    # @return [Integer]
    def day
      date_integer = @value[0, 2].to_i
      return @value[0, 2].to_i if date_integer < 32
      if (date_integer > 40) && (date_integer < 71)
        return @value[0, 2].to_i - 40
      end
      nil
    end

    # Get a numeric representation of the month of birth or registration
    #
    # @return [Integer]
    def month
      @value[2, 2].to_i
    end

    # Get the age of entity in years. Useful when dealing with age restrictions.
    #
    # @return [Integer]
    #
    # @example Get the current age of the entity
    #          k = Kennitala.new('0101302989')
    #          => #<Kennitala:0x007fe35d041bc0 @value="0101302989">
    #          k.age
    #          => 86
    def age
      year_diff = Date.today.year - to_date.year
      month_diff = Date.today.month - to_date.month
      day_diff = Date.today.month - to_date.month

      if month_diff.negative? || (month_diff.zero? && day_diff.zero?)
        return year_diff -= 1
      end
      year_diff
    end

    # Cast the kennitala to a Date object
    #
    # @return [Date]
    #
    # @example Cast the kennitala to a Date object
    #          k = Kennitala.new('0101302989')
    #          => #<Kennitala:0x007fe35d041bc0 @value="0101302989">
    #          k.to_date
    #          => #<Date: 1930-01-01 ((2425978j,0s,0n),+0s,2299161j)>
    def to_date
      Date.new(year, month, day)
    end

    # Cast the kennitala to a String object
    #
    # @return [String]
    def to_s
      @value.to_s
    end

    # Pretty print a kennitala
    #
    # Adds a space between the 6th and the 7th digits for readability.
    #
    # @param [String] spacer A single space by default
    # @return [String]
    #
    # @example Pretty print a Kennitala it its default form
    #          k = Kennitala.new('0101302989')
    #          => #<Kennitala:0x007fc589339f18 @value="2009155509">
    #          k.pp
    #          => "010130 2989"
    #
    # @example Use a hyphen as a spacer instead of the default single space
    #          k = Kennitala.new('0101302989')
    #          => #<Kennitala:0x007fc589339f18 @value="2009155509">
    #          k.pp('-')
    #          => "010130-2989"
    def pp(spacer = ' ')
      unless spacer.class == String
        raise ArgumentError 'Spacer must be a string'
      end
      @value[0, 6] + spacer + @value[6, 9]
    end

    private

    # Generate fake a birth number and check digit based on the first 6 digits
    #
    # @param [String] date_hash The first six digits is a kennitala
    # @return [Hash, nil]
    def fake_randoms(date_hash)
      first_six = date_hash[:day] + date_hash[:month] + date_hash[:year]
      loop do
        birth_number = Random.rand(1..99).to_s.rjust(2, '0')
        first_eight = "#{first_six}#{birth_number}"
        check_digit = calculate_check_digit(first_eight)
        if check_checksum(first_eight)
          return { check_digit: check_digit, birth_number: birth_number }
        end
      end
    end

    # Generate a fake year and century Hash
    #
    # @return [Hash]
    def fake_year
      century = [9, 9, 9, 8, 0, 0].sample
      current_year = Date.today.strftime('%y').to_i
      if century.zero?
        return { year: Random.rand(0..current_year), century: century }
      end
      { year: Random.rand(0..99), century: century }
    end

    # Generate a fake hash that includes randomly generated date elements
    #
    # @todo Take leap years into account and thus allow for 29 February
    #
    # @param [Boolean] is_company true if the day string is for a company
    # @return [Hash]
    def fake_date_hash(is_company = false)
      year_hash = fake_year

      month_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
      month = Random.rand(1..12)

      day = Random.rand(1..month_days[month - 1])
      day = (day.to_i + 40).to_s.rjust(2, '0') if is_company == true

      { century: year_hash[:century].to_s,
        year: year_hash[:year].to_s.rjust(2, '0'),
        month: month.to_s.rjust(2, '0'), day: day.to_s.rjust(2, '0') }
    end

    # Generate a fake, random kennitala string
    #
    # @param [Boolean] is_company If the kennitala is for a company/org
    # @return [String] A 10-digit string representing a kennitala
    def fake_kt_string(is_company = false)
      date_hash = fake_date_hash(is_company)
      randoms_hash = fake_randoms(date_hash)

      first_six = date_hash[:day] + date_hash[:month] + date_hash[:year]
      randoms = randoms_hash[:birth_number].to_s +
                randoms_hash[:check_digit].to_s

      first_six + randoms + date_hash[:century]
    end

    # Get year from a kennitala string
    #
    # @param [String] kt_string Sanitized kennitala string
    # @return [Integer] description of returned object
    def get_year_from_string(kt_string)
      century_code = kt_string[9, 1].to_i
      case century_code
      when 0
        return "20#{kt_string[4, 2]}".to_i
      when 8..9
        return "1#{century_code}#{kt_string[4, 2]}".to_i
      end
    end

    # Sanitize the kennitala
    #
    # @param [String] kt_string Unsanitised string representing a kennitala
    # @return [String, nil] Sanitized kennitala, nil if invalid
    def sanitize(kt_string)
      sanitized_kt = kt_string.gsub(/[^0-9]/, '')
      checks = check_checksum(sanitized_kt)

      year = get_year_from_string(sanitized_kt)
      day = sanitized_kt[0, 2].to_i
      day -= 40 if day > 40
      month = sanitized_kt[2, 2].to_i
      date = Date.new(year, month, day)

      return sanitized_kt if checks == true && date.class == Date
    rescue ArgumentError
      nil
    end

    # Calculate the checksum
    #
    # @param [String] kt_string Sanitized kennitala
    # @return [Integer] The checksum
    def checksum(kt_string)
      checksum = 0
      multipliers = [3, 2, 7, 6, 5, 4, 3, 2]
      multipliers.each_with_index do |multiplier, index|
        checksum += multiplier * kt_string[index].to_i
      end
      checksum
    end

    # Calculate the check digit for a kennitala
    #
    # @param [String] kt_string Sanitized kennitala
    # @return [Integer]
    def calculate_check_digit(kt_string)
      remainder = checksum(kt_string).modulo(11)

      # A kennitala with a remainder of 10 is always considered to be invalid
      return nil if remainder == 10

      # The check digit should be 11 minus the remainder,
      # unless the remainder is 0, then the theck digit becomes 0.
      return 0 if remainder.zero?
      11 - remainder
    end

    # Check validity of the check digit
    #
    # @param [String] kt_string Sanitized kennitala
    # @return [Boolean, nil] true on success false if the check digit is invalid
    def check_checksum(kt_string)
      expected_check_digit = calculate_check_digit(kt_string)
      actual_check_digit = kt_string[8].to_i
      return true if expected_check_digit == actual_check_digit
      false
    end
  end
end
